### DISCUSSION TITLE
Allow usage of PackageLoaders for Jinja2Templates

### DISCUSSION THREAD

**Discussion by DavidVentura** [encode/starlette#2235] [Ideas] [answered]
Hi
I deploy all my apps as [shiv](https://shiv.readthedocs.io/en/latest/) bundles; which is a way to deploy code and static content as a single file.
To do that, I serve the static files out of a [PackageLoader](https://jinja.palletsprojects.com/en/3.0.x/api/#jinja2.PackageLoader).
This functionality is not possible to use right now with `starlette`, as `Jina2Templates` [forces](https://github.com/encode/starlette/blob/master/starlette/templating.py#L127C9-L127C52) usage of a `FileSystemLoader`.

By patching `Jinja2Templates._create_env` to take a `loader`, I can use it like this:

```py
templates = CustomJinja2Templates(jinja2.PackageLoader('my_package.static', 'templates'))
```

My current overload (based on an older version of `starlette`) looks like this:
```py
import typing

import jinja2

from fastapi.templating import Jinja2Templates
from starlette.templating import pass_context


class CustomJinja2Templates(Jinja2Templates):
    # This class is a small override on
    # https://github.com/encode/starlette/blob/master/starlette/templating.py
    # to remove the need for a FileSystemLoader
    def __init__(self, loader: jinja2.PackageLoader):
        super().__init__(loader)

    def _create_env(
            self, loader: jinja2.PackageLoader, **env_options: typing.Any
    ) -> "jinja2.Environment":
        @pass_context
        def url_for(context: dict, name: str, **path_params: typing.Any) -> str:
            request = context["request"]
            return request.url_for(name, **path_params)

        env_options.setdefault("loader", loader)
        env_options.setdefault("autoescape", True)

        env = jinja2.Environment(**env_options)
        env.globals["url_for"] = url_for
        return env
```

The only real difference is that `_create_env` takes a `loader` argument.

**Comment by alex-oleshkevich** [comment, marked as answer]
You can pass a custom Jinja loader to the templates - https://www.starlette.io/templates/#using-custom-jinja2environment-instance
